pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/DateTruncTransformFunction.java

```diff
@@ -0,0 +1,261 @@
+  private static final String UTC_TZ = TimeZoneKey.UTC_KEY.getId();
+  private static final DateTimeFieldType QUARTER_OF_YEAR = new QuarterOfYearDateTimeField();
+  private static final Logger LOGGER = LoggerFactory.getLogger(DateTruncTransformFunction.class);
+  private TransformFunction _mainTransformFunction;
+  private TransformResultMetadata _resultMetadata;
+  private long[] _longOutputTimes;
+  private DateTimeField _field;
+  private TimeUnit _inputTimeUnit;
+  private TimeUnit _outputTimeUnit;
+
+  private static DateTimeField getTimestampField(ISOChronology chronology, String unitString) {
+    switch (unitString) {
+      case "millisecond":
+        return chronology.millisOfSecond();
+      case "second":
+        return chronology.secondOfMinute();
+      case "minute":
+        return chronology.minuteOfHour();
+      case "hour":
+        return chronology.hourOfDay();
+      case "day":
+        return chronology.dayOfMonth();
+      case "week":
+        return chronology.weekOfWeekyear();
+      case "month":
+        return chronology.monthOfYear();
+      case "quarter":
+        return QUARTER_OF_YEAR.getField(chronology);
+      case "year":
+        return chronology.year();
+    }
+    throw new IllegalArgumentException("'" + unitString + "' is not a valid Timestamp field");
+  }
+
+  @Override
+  public String getName() {
+    return FUNCTION_NAME;
+  }
+
+  @Override
+  public void init(List<TransformFunction> arguments, Map<String, DataSource> dataSourceMap) {
+    if (arguments.size() < 3) {
+      throw new IllegalArgumentException(
+          String.format("At least 3 arguments are required for %s transform function", FUNCTION_NAME));
+    }
+    String unit = ((LiteralTransformFunction) arguments.get(0)).getLiteral().toLowerCase();
+    TransformFunction valueArgument = arguments.get(1);
+    if (valueArgument instanceof LiteralTransformFunction || !valueArgument.getResultMetadata().isSingleValue()) {
+      throw new IllegalArgumentException(
+          "The second argument of dateTrunc transform function must be a single-valued column or a transform function");
+    }
+    _mainTransformFunction = valueArgument;
+    String inputTimeUnitStr = ((LiteralTransformFunction) arguments.get(2)).getLiteral().toUpperCase();
+    _inputTimeUnit = TimeUnit.valueOf(inputTimeUnitStr);
+
+    String timeZone = arguments.size() >= 4 ? ((LiteralTransformFunction) arguments.get(3)).getLiteral() : UTC_TZ;
+    String outputTimeUnitStr =
+        arguments.size() >= 5 ? ((LiteralTransformFunction) arguments.get(4)).getLiteral().toUpperCase()
+            : inputTimeUnitStr;
+    TimeZoneKey timeZoneKey = TimeZoneKey.getTimeZoneKey(timeZone);
+
+    _field = getTimestampField(DateTimeZoneIndex.getChronology(timeZoneKey), unit);
+    _resultMetadata = LONG_SV_NO_DICTIONARY_METADATA;
+    _outputTimeUnit = TimeUnit.valueOf(outputTimeUnitStr);
+
+    if (arguments.size() > 5) {
```

Comment:
This can be combined with the first check in line 130 `Preconditions.checkArgument(numArguments >= 3 && numArguments <= 5, "...")` and provide an example of the expected arguments in the exception message